src/apple/alohalytics_objc.mm

```diff
@@ -266,6 +265,58 @@ static void LogSystemInformation() {
   }
   return parsed;
 }
+
+// Need it to effectively upload data when app goes into background.
+static UIBackgroundTaskIdentifier sBackgroundTaskId = UIBackgroundTaskInvalid;
+static void EndBackgroundTask() {
+  [[UIApplication sharedApplication] endBackgroundTask:sBackgroundTaskId];
+  sBackgroundTaskId = UIBackgroundTaskInvalid;
+}
+static void OnUploadFinished(alohalytics::ProcessingResult result) {
+  if (Stats::Instance().DebugMode()) {
+    const char * str;
+    switch (result) {
+      case alohalytics::ProcessingResult::ENothingToProcess: str = "There is no data to upload."; break;
+      case alohalytics::ProcessingResult::EProcessedSuccessfully: str = "Data was uploaded successfully."; break;
+      case alohalytics::ProcessingResult::EProcessingError: str = "Error while uploading data."; break;
+    }
+    ALOG(str);
+  }
+  if (sBackgroundTaskId != UIBackgroundTaskInvalid) {
+    EndBackgroundTask();
+  }
+}
+
+// Quick check if device has any active connection.
+// Does not guarantee actual reachability of any host.
+bool IsConnectionActive() {
```

Comment:
This is pure black magic for most people except few in the Apple.
Main idea was taken from their recommended example app https://developer.apple.com/library/ios/samplecode/Reachability/Introduction/Intro.html <https://developer.apple.com/library/ios/samplecode/Reachability/Introduction/Intro.html>
> On Jun 10, 2015, at 19:52, Dima <notifications@github.com> wrote:
> 
> In src/apple/alohalytics_objc.mm <https://github.com/KnowSheet/Alohalytics/pull/36#discussion_r32139195>:
> 
> > +    const char * str;
> > +    switch (result) {
> > +      case alohalytics::ProcessingResult::ENothingToProcess: str = "There is no data to upload."; break;
> > +      case alohalytics::ProcessingResult::EProcessedSuccessfully: str = "Data was uploaded successfully."; break;
> > +      case alohalytics::ProcessingResult::EProcessingError: str = "Error while uploading data."; break;
> > +    }
> > +    ALOG(str);
> > +  }
> > +  if (sBackgroundTaskId != UIBackgroundTaskInvalid) {
> > +    EndBackgroundTask();
> > +  }
> > +}
> > +
> > +// Quick check if device has any active connection.
> > +// Does not guarantee actual reachability of any host.
> > +bool IsConnectionActive() {
> Note to self: This is black magic to me.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/KnowSheet/Alohalytics/pull/36/files#r32139195>.
>